modules/geo-layers/src/mvt-layer/mvt-layer.js

```diff
@@ -84,6 +80,110 @@ export default class MVTLayer extends TileLayer {
       feature => getFeatureUniqueId(feature, uniqueIdProperty) === featureIdToHighlight
     );
   }
+
+  updateState({props, oldProps, context, changeFlags}) {
+    super.updateState({props, oldProps, context, changeFlags});
+    this._debouncedGetViewportFeatures();
+  }
+
+  _getModelMatrixScale(tile) {
+    const worldScale = Math.pow(2, tile.z);
+    const xScale = WORLD_SIZE / worldScale;
+    const yScale = -xScale;
+
+    return [xScale, yScale, 1];
+  }
+
+  _getCoordinateOrigin(tile) {
+    const worldScale = Math.pow(2, tile.z);
+    const xOffset = (WORLD_SIZE * tile.x) / worldScale;
+    const yOffset = WORLD_SIZE * (1 - tile.y / worldScale);
+
+    return [xOffset, yOffset, 0];
+  }
+
+  _debouncedGetViewportFeatures() {
+    const {onViewportFeatures} = this.props;
+
+    if (onViewportFeatures) {
+      let {viewportChangeTimer} = this.state;
+      clearTimeout(viewportChangeTimer);
+
+      viewportChangeTimer = setTimeout(() => {
+        const {tileset} = this.state;
+        const {isLoaded} = tileset;
+
+        if (isLoaded) {
+          this._getViewportFeatures();
+        }
+      }, VIEWPORT_CHANGE_DEBOUNCE);
+
+      this.setState({viewportChangeTimer});
+    }
+  }
+
+  async _getViewportFeatures() {
+    const {tileset} = this.state;
+    const {onViewportFeatures} = this.props;
+    const featureCache = new Set();
+    const currentFrustumPlanes = this.context.viewport.getFrustumPlanes();
+    let viewportFeatures = [];
+
+    await tileset.selectedTiles.forEach(async tile => {
+      const features = (await tile.data) || [];
+      const transformationMatrix = new Matrix4()
+        .translate(this._getCoordinateOrigin(tile))
+        .scale(this._getModelMatrixScale(tile));
+
+      viewportFeatures = viewportFeatures.concat(
+        features.filter(f => {
+          const featureId = getFeatureUniqueId(f, this.props.uniqueIdProperty);
+          if (
+            !featureCache.has(featureId) &&
+            this._checkIfCoordinatesIsInsideFrustum(
+              transformationMatrix,
+              currentFrustumPlanes,
+              f.geometry.coordinates
+            )
+          ) {
+            featureCache.add(featureId);
+            return true;
+          }
+          return false;
+        })
+      );
+    });
+
+    onViewportFeatures(viewportFeatures);
+  }
+
+  _checkIfCoordinatesIsInsideFrustum(matrix, frustumPlanes, coordinates) {
+    if (Array.isArray(coordinates) && coordinates.length && typeof coordinates[0] === 'number') {
+      return this._coordinateInPlanes(frustumPlanes, matrix.transform(coordinates).concat(0));
+    }
+
+    return coordinates.some(c => {
```

Comment:
This is technically incorrect: a path can both start and end outside the viewport but still visible.

Reply:
Yes, you are right :(
As @ibgreen says, we can remove this code and use Deck.pickObjects instead, right?

Reply:
Pickobjects isn’t perfect either - for something to be picked it needs to cover at least 1 pixel, so there is a chance some object won’t be picked if overlapped etc. but this is very rare and in practice it works really well and performance makes it well worth it.

But if you need 100% guarantee the number is correct you may need to count.